aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -591,16 +639,25 @@ Tensor scaled_dot_product_attention(
       bool compute_logsumexp =
           (query_.requires_grad() || key.requires_grad() ||
            value.requires_grad());
+      if (attn_mask.has_value()) {
+        // Expand to 4d case
+        attn_mask = attn_mask.value().expand_symint(
+            {query_.sym_size(0),
+             query_.sym_size(1),
+             query_.sym_size(2),
+             key.sym_size(2)});
+        attn_mask = pad_bias(attn_mask.value());
```

Comment:
So when the attention mask is a bool, you'll have allocated a new float attention mask in `convert_boolean_attn_mask` and then subsequently will reallocate to pad it out to a larger one if it's not a multiple of 16. 

Seems like an opportunity to allocate the right size right away if you need padding, i.e. extend `convert_boolean_attn_mask` to also do the padding. You'll know at `convert_boolean_attn_mask` time what backend you're about to choose, so it doesn't require more special casing.